server/tso/global_allocator.go

```diff
@@ -202,99 +301,96 @@ func (gta *GlobalTSOAllocator) SyncMaxTS(ctx context.Context, dcLocationMap map[
 				}
-			}(ctx, leaderConn, respCh, errCh)
+			}(ctx, leaderConn, respCh)
 		}
 		wg.Wait()
 		close(respCh)
-		close(errCh)
-		// If any error occurs, the synchronization process will fail
-		if err := <-errCh; err != nil {
-			errList = append(errList, err)
-		}
-		if len(errList) > 0 {
-			return errs.ErrSyncMaxTS.FastGenWithCause(errList)
-		}
 		var (
-			respCount         int
-			syncedDCs         []string
-			inCollectingPhase bool
+			errList   []error
+			syncedDCs []string
+			maxTSORtt time.Duration
 		)
+		// Iterate each response to handle the error and compare MaxTSO.
 		for resp := range respCh {
-			respCount++
-			if resp == nil {
-				return errs.ErrSyncMaxTS.FastGenByArgs("got nil response")
+			if resp.err != nil {
+				errList = append(errList, resp.err)
 			}
-			// Once we get a non-nil and non-zero MaxLocalTs first, we will think it's in the first phase
-			// of the Global TSO synchronization. So that we can have more detailed processing logic
-			// for each phase. For example, if we think we're in the first phase of the Global TSO
-			// synchronization, the inCollectingPhase will be set to true, and during this phase,
-			// any response with nil or empty MaxLocalTs will be regarded as an invalid response.
-			// Then the whole synchronization will fail.
-			if respCount == 1 && resp.GetMaxLocalTs() != nil && resp.GetMaxLocalTs().GetPhysical() != 0 {
-				inCollectingPhase = true
+			// If any error occurs, just jump out of the loop.
+			if len(errList) != 0 {
+				break
 			}
-			if inCollectingPhase {
-				// Handle the response of the first phase: collect all the Local TSOs
-				if resp.GetMaxLocalTs() == nil || resp.GetMaxLocalTs().GetPhysical() == 0 {
-					return errs.ErrSyncMaxTS.FastGenByArgs("got nil or zero max local ts in the first sync phase")
-				}
-				// Compare and get the max one
-				if tsoutil.CompareTimestamp(resp.GetMaxLocalTs(), maxTSO) > 0 {
-					*maxTSO = *(resp.GetMaxLocalTs())
+			if resp.rpcRes == nil {
+				return errs.ErrSyncMaxTS.FastGenByArgs("got nil response")
+			}
+			if skipCheck {
+				// Set all the Local TSOs to the maxTSO unconditionally, so the MaxLocalTS in response should be nil.
+				if resp.rpcRes.GetMaxLocalTs() != nil {
+					return errs.ErrSyncMaxTS.FastGenByArgs("got non-nil max local ts in the second sync phase")
 				}
-				syncedDCs = append(syncedDCs, resp.GetDcs()...)
+				syncedDCs = append(syncedDCs, resp.rpcRes.GetSyncedDcs()...)
 			} else {
-				// Handle the response of the second phase: set all the Local TSOs to the maxTSO
-				if resp.GetMaxLocalTs() != nil {
-					return errs.ErrSyncMaxTS.FastGenByArgs("got non-nil max local ts in the second sync phase")
+				// Compare and get the max one
+				if tsoutil.CompareTimestamp(resp.rpcRes.GetMaxLocalTs(), maxTSO) > 0 {
+					*maxTSO = *(resp.rpcRes.GetMaxLocalTs())
+					if resp.rtt > maxTSORtt {
+						maxTSORtt = resp.rtt
+					}
 				}
```

Comment:
What about following suggestion?

```suggestion
				// Compare and get the max one
				if tsoutil.CompareTimestamp(resp.rpcRes.GetMaxLocalTs(), maxTSO) > 0 {
					*maxTSO = *(resp.rpcRes.GetMaxLocalTs())
				}
				if resp.rtt > maxTSORtt {
				        maxTSORtt = resp.rtt
				}
```

Reply:
I think we should only consider the RTT of request which has a bigger Local TS than `MaxTSO` here, otherwise, the RTT may contain the potential etcd writing time.

Reply:
got it.